### PR TITLE
SubscriberStatelessServiceBase

### PR DESCRIPTION
**Pull request by danadesrosiers**
Added SubscriberStatelessServiceBase class to simplify managing Stateless Service subscribers.

Abstracts boilerplate work of subscribing to messages and exposing endpoints for subscribing and unsubscribing.

Provides a simple mechanism to accept and handle multiple message types.

### CODE REVIEW

**Comment by loekd** [comment]
Thanks for this contribution. I've changed your code a bit to use the PayloadSerializer.
Also thought of a way to make subscriptions declarative, instead of imperative.
Please let me know your thoughts on this. We can do the same for stateful services.

File: ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs
Comment (danadesrosiers) [inline ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs]: I see why the async/await wasn't necessary.  We should update UnsubscribeAsync() as well.

File: ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs
Comment (danadesrosiers) [inline ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs]: We should also add a parameter for the broker Uri.

**Review by danadesrosiers — COMMENTED** [review]
It looks good, I like the changes.
Serializer changes are nice.
I'm really happy to see the addition of a test project.
The annotation is a great idea.  I did not love the way types were being registered before.  I'm thinking it might make more sense to put the annotation on the handler method itself instead of at the class level.

File: ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs
Comment (loekd) [reply ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs]: Agreed, made some changes accordingly.
Also, I'm thinking we could add a bootstrapper to the broker service, that would detect subscriber services and automatically register them. Potentially, we can avoid introduction of a required service base class, making the library more applicable in projects. Still struggling with how to discover the service CLR type while potentially running inside a different application. 

Maybe we just add a `Register` operation to `ISubscriberService` and invoke that?

Comment (danadesrosiers) [reply ServiceFabric.PubSubActors/SubscriberServices/SubscriberStatelessServiceBase.cs]: The bootstrapper is an interesting idea.  If the Broker can automatically discover and register subscribers, it could dramatically simplify the consumer.  Being able to work with a CLR type that is not present in the application is a tough problem.  I'm not sure it's possible.  I was not able to put much time into this today, but I will see if I can figure something out tomorrow.

**Comment by danadesrosiers** [comment]
My attempts at implementing the SubscriberServiceBootstrap did not go well.  As far as I can see, we don't have any good way to know if a given service is a subscriber service.  The ServiceProxy itself doesn't know anything about the service, it was created based on an interface and location that the user provides and doesn't necessarily match the actual service.  It might be possible if we have an operation on ISubscriberService to return the service's interface type which would define handler methods annotated with Subscribe attributes.  There are more details to figure out, but it would definitely require that the Broker has knowledge of all the service interfaces and event types, which I would really rather not do.

Also, I was thinking about ISubscriberStatelessService.  I created it for use with SubscriberStatelessServiceBase that added the SubscribeAsync and UnsubscribeAsync endpoints.  I'm wondering, should these be added to ISubscriberService so all subscribers have these endpoints, or perhaps we don't need these at all?  They are a pass-through to the Broker, so maybe it is possible to call the Broker service directly.  What do you think?

**Comment by loekd** [comment]
About (1), we route messages by type name, so we don't need the CLR types themselves. Simply returning a list of strings would suffice. 
As an alternative, we could consider using [labels](https://docs.traefik.io/configuration/backends/servicefabric/), like Traefik does.

About (2), a subscriber really only needs a method to receive messages. I'd be happier if we either add metadata to a subscriber service and have it automagically registered or use the current way of having the service register itself. (It's not required for a service to register itself, it can be done on behalf of a service, so we don't need an endpoint to order it to subscribe itself. This feels unnatural.)

**Comment by koderi-dp** [comment]
Why not go with property manager from sf, like in broker, for subsriber to register itself? Meta can be serilized json but keep it small, then it's easy to discover subsriber service from broker service, also you clould put inside meta supported message types? Or base subsriber interface could also have something like getsupportedmessages which can be called by broker to register it for those messages

Like the idea

**Comment by danadesrosiers** [comment]
My point is that the Broker cannot know anything about the subscriber service without asking it.  Labels would probably work, but the opinion over here is that would not be a very intuitive experience for the developer.  It also may not be flexible enough for all use cases.  The complex situation we are thinking of is when a Stateful subscriber service wants to choose whether or not it wants to subscribe to a message type based on which partition it is on.

I'm not sure that having the Broker look for subscriber services simplifies anything.  The Broker still needs to call to the subscriber to get subscription information.  The subscriber still needs to set up handlers and gather information about which message types it wants to subscribe to.  On a positive note, I was able to move the handler discovery logic to the SubscriberServiceHelper, so it will be reusable in Stateful subscriber services.

I definitely agree that it feels unnatural to have an endpoint that just passes the subscribe or unsubscribe request through to the Broker when we could just call the Broker service directly.  Currently, the RegisterServiceSubscriberAsync endpoint takes a ServiceReference object.  Will the Broker be able to detect all the things it needs to build the ServiceReference properly (i.e. partition, listener)?  Will it be challenging to craft the ServiceReference consistently when sometimes it is registered by the Broker and sometimes the request comes from the subscriber?  If we are not careful with the comparison logic, we may end up with duplicate subscriptions.  Probably the new ISubscriberService endpoint would have to return the ServiceReference along with the list of message types.  Then we have just created a bulk subscribe action, which is not bad, but I'm not sure it gets us much.  Whether that action is initiated by the Broker or the Subscriber service doesn't seem to make much much difference.

**Comment by danadesrosiers** [comment]
Doing some more investigation, I found that ServiceFabric properties can be created dynamically, so maybe there is still hope for that idea.

**Comment by loekd** [comment]
Would you mind elaborating a little on the dynamic properties?

**Comment by danadesrosiers** [comment]
Sorry for slow response, I'm out of town for a few days.  

I was able to use https://docs.microsoft.com/en-us/dotnet/api/system.fabric.fabricclient.propertymanagementclient.putpropertyasync?view=azure-dotnet#System_Fabric_FabricClient_PropertyManagementClient_PutPropertyAsync_System_Uri_System_String_System_String_ to add a property on the subscriber Service Fabric service from the OnOpenAsync() call and read that property from the Broker, also using the PropertyManagementClient.

Potentially, the subscriber service could build some meta data about itself, probably mostly drawing from the Subscribe attribute, and store that meta data in a property.  The broker could read that property to know if the service is a subscriber and which message types to subscribe to.

I'm not sure if there is a notification that can be triggered to alert the Broker that a property has changed.  If not, that could be a problem.

**Comment by loekd** [comment]
I've thought of a way to auto register subscribers without adding much complexity. Check the latest push. I think that could work. Any thoughts?

**Comment by danadesrosiers** [comment]
You have used the bootstrapper as a factory to create the Service instance and register subscribers in one step.  If you could create an easy to use factory that would set everything up and not need a base service class anymore, then I think this would be an interesting approach.  But I don't think that is going to work out.  With this approach, we still need to get the collection of SubscriptionDefinitions to the Service and implement ReceiveMessageAsync() to call the appropriate handler.

This seems more complex than subscribing in OnOpenAsync() since the implementer needs to do setup work in two places now, Program and the Service class.  Did you see my previous commit where I moved most of the logic into SubscriberServiceHelper?  A lot of it got reverted in your "merge down master" commit.  We can move all the subscription logic into SubscriberServiceHelper so it can be reused for a SubscriberStatefulServiceBase.  That should be pretty simple to use.

**Comment by danadesrosiers** [comment]
See my latest commit.  I moved all logic into the SubscriberServiceHelper.  I currently have the SubscriptionDefinitions stored in the helper.  I'm not sure if I like that.  It may be better to store that in the base class so the helper doesn't need to maintain any state.  I also played around with an idea to have the caller send a ServiceReference object instead of itself to minimize having to create two versions of everything in order to support Stateless and Stateful services.  Let me know what you think.

**Comment by loekd** [comment]
I propose we separate the responsibilities of subscribing and method invocation.

Using the factory delegate doesn't require using a base class, nor any implementation on the end-user side, it only requires a Stateless or Stateful service from the SF SDK and the interface `ISubscriberService`. The factory class will do all the hard work of discovering attributes on the service instance and register subscriptions. The logic of invoking the correct method when a message arrives can be separated from the process of subscribing, by introducing a helper. You'd keep the subscription defs in memory as part of the service composition, but not its base. The helper would be injected by the bootstrapper.

We can delete the `SubscriberStatefulServiceBase` class and only require a opt-in or opt-out mechanisms for auto-registration and auto-invocation. E.g. by using a service class attribute.

My apologies for the destructive merge, that was unintended and messy.

**Comment by danadesrosiers** [comment]
So, usage for the bootstrapper approach would look something like this:
```
public class MySubscriberStatelessService : StatelessService, ISubscriberService
{
    private readonly ISubscriberServiceHelper _subscriberServiceHelper;

    protected SubscriberStatelessServiceBase(StatelessServiceContext serviceContext, ISubscriberServiceHelper subscriberServiceHelper)
            : base(serviceContext, subscriberServiceHelper)
    {
    }

    public Task ReceiveMessageAsync(MessageWrapper messageWrapper)
    {
        return _subscriberServiceHelper.ProccessMessageAsync(messageWrapper);
    }

    protected override IEnumerable<ServiceInstanceListener> CreateServiceInstanceListeners()
    {
        return this.CreateServiceRemotingInstanceListeners();
    }

    [Subscribe(typeof(TestMessage))]
    private Task HandleTestMessage(TestMessage testMessage)
    {
        // TODO: handle message
    }
}
```
This assumes the bootstrapper somehow made the helper aware of the SubscriptionDefinitions so SubscriberServiceHelper.ProcessMessageAsync() is able to route the message to the proper handler.

Then in Program.cs, we have to modify the service registration to something like this:
```
ServiceRuntime.RegisterServiceAsync("MyApp.MySubscriberServiceType",
    context => new StatelessSubscriberServiceBootstrapper<MySubscriberStatelessService >(context, ctx => new MySubscriberStatelessService (ctx, new SubscriberServiceHelper())).Build())
    .GetAwaiter().GetResult();
```

Using the base class, it would be something like this:
```
public class MySubscriberStatelessService : SubscriberStatelessService
{
    protected SubscriberStatelessServiceBase(StatelessServiceContext serviceContext)
            : base(serviceContext)
    {
    }

    [Subscribe(typeof(TestMessage))]
    private Task HandleTestMessage(TestMessage testMessage)
    {
        // TODO: handle message
    }
}
```
And service registration is simply the generated default.  My goal is to make this a simple as possible to use.  In this case, the implementer doesn't need to deal with ISubscriberServiceHelper directly.  Boilerplate setup is handled automatically.

Can you explain why you would like to avoid having a base class?  It really helps simplify usage.  Yes, we would need a separate base for Stateful and Stateless, but all the logic is in the helper, so that's not a big deal.  The bootstrapper needs two implementations as well.  The only issue I can think of is if the service needs to implement another base class for some reason.  That seems like it would be rare and the solution would be to use the helper to manually subscribe and process messages.  That worst case situation would be only one additional method call more than the best case of the bootstrap approach.

I think having the ability opt-out of subscribing, as you mentioned, is useful, but probably rare.  Maybe an attribute on the service class like `[AutoSubscribe(false)]`.  A StatefulService may use such a feature to decide whether or not to subscribe based on which partition it's on.  I don't think an opt-out for auto invocation makes sense because the user explicitly opts in by using the `Subscribe` attribute.  Besides, it doesn't hurt to register handlers that will never be invoked because we never subscribed to that type.

**Comment by loekd** [comment]
That's a solid argument. I favor composition over inheritance to allow maximum flexibility at the end-user side, but indeed this comes at the price of adding some complexity. So why don't we meet in the middle, and offer both the base class (using the helper) and the bootstrapper + helper? It doesn't require a lot more than what is here now.

So it would become:

```
public class MySubscriberStatelessService : SubscriberStatelessService
{
    protected SubscriberStatelessServiceBase(StatelessServiceContext serviceContext, ISubscriberServiceHelper subscriberServiceHelper = null)
            : base(serviceContext, subscriberServiceHelper)
    {
    }

    [Subscribe(typeof(TestMessage))]
    private Task HandleTestMessage(TestMessage testMessage)
    {
        // TODO: handle message
    }
}
```

Passing `ISubscriberServiceHelper` to the ctor to allow mocking and unit testing.

**Comment by danadesrosiers** [comment]
In the SubscriberStatelessServiceBase, the ISubscriberServiceHelper is an optional argument in the constructor, so implementers have it available if they need it for testing or anything else, but don't have to provide it if they don't need it.

I agree, composition is preferable over inheritance when possible.  I think we achieved that goal by moving all logic into the helper.  The base class provides a convenient wrapper to abstract away the plumbing, but it is not required.  It is easy enough to use the helper directly instead of the base class when maximum flexibility is required.  In that case, I think it is more simple to subscribe in the service manually rather than setting up a bootstrapper to do it automatically.  At least everything would be in one place.

I tested the SubscriberServiceBootstrapper (with a couple modifications to get it working) and ran into a problem.  It turns out, at the time the service is instantiated, the "Partition" property on the service is not set yet.  When we try to create the ServiceReference so we can subscribe, we get a null reference exception.  We could do handler discovery in the bootstrapper, but it looks like subscribing will have to happen in OnOpenAsync().

**Comment by danadesrosiers** [comment]
@loekd, looking forward to your feedback on the latest changes.  We are eager to wrap up this task.

**Review by loekd — APPROVED** [review]

**Comment by loekd** [comment]
I've added some code, and the bootstrapper now works. I found out that the `SubscriberStatelessServiceBase` doesn't unregister itself when it moves, or is shutdown at this time.
Once that is in place, we can merge this.

**Comment by danadesrosiers** [comment]
We are intentionally not unregistering.  If a service moves or goes down for any reason, we want the Broker to continue to queue up incoming messages and deliver them when the service becomes available again so we don't lose any messages.  We are thinking services should be manually unregistered via some Broker management console.

**Comment by loekd** [comment]
Alright, makes sense.
I've pushed my final changes. When you agree, I'll push the merge button.

**Comment by danadesrosiers** [comment]
I added a base class for stateful services and updated the documentation accordingly.  I'm ready to merge.

**Comment by loekd** [comment]
The new version is released. Thanks @danadesrosiers  for all the hard work. I really like these improvements.

**Pull request merged**
